pulsar-io/kafka-connect-adaptor/src/main/java/org/apache/pulsar/io/kafka/connect/KafkaSinkWrappingProducer.java

```diff
@@ -0,0 +1,343 @@
+        sinkContext.throwIfNeeded();
+        CompletableFuture<RecordMetadata> result = new CompletableFuture<>();
+        try {
+            task.put(Lists.newArrayList(toSinkRecord(producerRecord)));
+        } catch (Exception ex) {
+            result.completeExceptionally(ex);
+            return result;
+        }
+        pendingFlush.whenComplete((ignore, ex) -> {
+            if (ex == null) {
+                result.complete(null);
+            } else {
+                result.completeExceptionally(ex);
+            }
+        });
+        numPendingRecords.incrementAndGet();
+        flushIfNeeded(false);
+        return result;
+    }
+
+    @Override
+    public Future<RecordMetadata> send(ProducerRecord<K, V> producerRecord, Callback callback) {
+        sinkContext.throwIfNeeded();
+        CompletableFuture<RecordMetadata> result = new CompletableFuture<>();
+        result.whenComplete((ignore, ex) -> {
+            if (ex == null) {
+                callback.onCompletion(null, null);
+            } else {
+                if (ex instanceof Exception) {
+                    callback.onCompletion(null, (Exception) ex);
+                } else {
+                    callback.onCompletion(null, new Exception(ex));
+                }
+            }
+        });
+
+        try {
+            task.put(Lists.newArrayList(toSinkRecord(producerRecord)));
+        } catch (Exception ex) {
+            result.completeExceptionally(ex);
+            return result;
+        }
+        pendingFlush.whenComplete((ignore, ex) -> {
+            if (ex == null) {
+                result.complete(null);
+            } else {
+                result.completeExceptionally(ex);
+            }
+        });
+        numPendingRecords.incrementAndGet();
+        flushIfNeeded(false);
+        return result;
+    }
+
+    private void flushIfNeeded(boolean force) {
+        if (force || numPendingRecords.get() >= batchSize) {
+            flush();
+        }
+    }
+
+    @Override
+    public void flush() {
+        if (log.isDebugEnabled()) {
+            log.debug("flush requested, pending: {}, batchSize: {}",
+                    numPendingRecords.get(), batchSize);
+        }
+
+        sinkContext.throwIfNeeded();
+
+        if (numPendingRecords.getAndSet(0) == 0) {
+            return;
+        }
+
+        Map<TopicPartition, OffsetAndMetadata> currentOffsets = taskContext.currentOffsets();
+        CompletableFuture<Void> flushCf;
+        synchronized (this) {
+            flushCf = pendingFlush;
+            pendingFlush = new CompletableFuture<>();
+        }
+
+        try {
+            task.flush(currentOffsets);
+            taskContext.flushOffsets(currentOffsets);
+            flushCf.complete(null);
+        } catch (Throwable t) {
+            log.error("error flushing pending records", t);
+            flushCf.completeExceptionally(t);
+        }
+    }
+
+    @Override
+    public List<PartitionInfo> partitionsFor(String topic) {
```

Comment:
do we need to implement this method ?
if we are not calling it from our KafkaSink than we could simply throw UnsupportedOperationException